abi.html

```diff
@@ -2807,66 +2898,90 @@ <h4><a href="#return-value"> 3.1.4 Return Values </a></h4>
 cause it to point anywhere, including to global or heap-allocated memory.
+
 <p>
+C ABIs usually provide treatment for "indirect" return values, e.g. when
+returning a large aggregate that cannot fit in registers.  In some cases,
+this treatment may not be suitable for non-trivial C++ return values, such
+as if the convention requires implicit copying or does not permit the
+return value to be constructed at an arbitrary address.  If the treatment
+exists and is suitable, it is used for non-trivial return values.
+Otherwise, the pointer is passed as if it were the first parameter in
+the function prototype, preceding all other parameters, including the
+<code>this</code> and <code>VTT</code> parameters.
 
 <p>
-If the C ABI uses a special rule for "indirect" return values, e.g. when
-the return type is too large to fit in registers, that rule is used to
-pass the pointer.  Otherwise the pointer is passed as the first parameter,
-preceding all other parameters (including the <code>this</code> and
-<code>VTT</code> parameters).
-</p>
+<a name="special-class-return-value">
+<h4><a href="#special-class-return-value"> 3.2.2 Return Values of Special Class Type </a></h4>
 
 <p>
-Reference return values are returned as if pointers to the object bound
-to the reference.
-</p>
+A return value of class
+<code>std::decimal::decimal32</code>,
+<code>std::decimal::decimal64</code>, or
+<code>std::decimal::decimal128</code> as defined in TR 24733 is
+returned the same as the corresponding native decimal floating-point
+scalar type.
 
 <p>
- Another exception is that a return value type of class
- <code>std::decimal::decimal32</code>,
- <code>std::decimal::decimal64</code>, or
- <code>std::decimal::decimal128</code> as defined in TR 24733 is
- returned the same as the corresponding native decimal floating-point
- scalar type.
-</p>
+<a name="reference-return-value">
+<h4><a href="#reference-return-value"> 3.2.4 Reference Return Values </a></h4>
 
 <p>
-A result of an empty class type will be returned as though it were
-a struct containing a single char,
-i.e. <code>struct S { char c; };</code>.
-The actual content of the return register is unspecified.
-On Itanium, the associated NaT bit must not be set.
+A return value of reference type is returned as a pointer to the object
+bound to the reference.
+
+<p>
+<a name="empty-return-value">
+<h4><a href="#emptty-return-value"> 3.2.4 Empty Return Values </a></h4>
 
+<p>
+If the base ABI specifies rules for returning an empty C <code>struct</code>,
+those rules apply when returning an empty class.  Otherwise empty classes
+are returned as if returning the corresponding C <code>struct</code> as
+described above for <a href="#empty-parameter">empty parameters</a>.
+The contents of any registers or memory associated with the return value
+are unspecified, and callers must not rely on receiving any particular
+value.  On Itanium, the NaT bit must not be set on all registers
+associated with the return value.
```

Comment:
Again this seems like the wrong suggestion for new ABIs. If the return type is trivial for calls and empty we should suggest behaving as we would for a void return.